gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use codec::{Decode, Encode};
+
+/// Constant defines codec index of [`Request::MultiMillerLoop`].
+pub const REQUEST_MULTI_MILLER_LOOP: u8 = 0;
+/// Constant defines codec index of [`Request::FinalExponentiation`].
+pub const REQUEST_FINAL_EXPONENTIATION: u8 = 1;
+/// Constant defines codec index of [`Request::MultiScalarMultiplicationG1`].
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G1: u8 = 2;
+/// Constant defines codec index of [`Request::MultiScalarMultiplicationG2`].
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G2: u8 = 3;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG1`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G1: u8 = 4;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG2`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G2: u8 = 5;
+
+/// Type that should be used to create a message to the bls12_381 builtin actor.
+#[derive(Encode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = codec)]
+pub enum Request {
+    /// Request to pairing multi Miller loop for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `a`: `ArkScale<Vec<G1Affine>>`.
```

Comment:
And reference crate inside hyperlink. Maybe consider types re-import
```suggestion
    ///   - `a`: [`ArkScale<Vec<G1Affine>>`].
```